Enable CLI builds with Eclipse Tycho

manandbytes opened the pull request.

manandbytes committed:
Default target platform

manandbytes committed:
Build with Tycho

manandbytes committed:
Enable Travis CI builds

manandbytes committed:
Build product

Don't use org.eclipse.core.runtime.products extension point as it
duplicates .product file.

Relocate icons and splash to the same directory as a product file and
reference them using relative path.

Drop custom config.ini file: it will be generated based on the .product
file.

Start org.eclipse.core.runtime bundle automatically.

All packaging is done by Tycho itself which makes package.sh script
obsolete.

nawroth commented:
@manandbytes This looks great! Didn't know Tycho was up to this (again).

Some input:

To merge the PR, we need a CLA, see: http://docs.neo4j.org/chunked/snapshot/cla.html

The packaging needs some work:
- The zips should have a root folder, just like Eclipse distributions use to have (but in this case, it should be named `neoclipse-{version}`).
- The version should be in the name of the zips too.
- The license, notice and readme files from inside `org.neo4j.neoclipse/` has to be included.

We do have the `org.neo4j.neoclipse/package.sh` script which could be modified. Or we could just use a different mvn plugin for the packaging part. Doesn't look like the one in use offers a lot of packaging options.

We shouldn't build distributions for platforms where we have no intention on testing them. I'd say, stick to:
- win x86
- win x86_64
- linux x86
- linux x86_64
- osx cocoa x86
- osx cocoa x86_64
- hpux

If someone steps up to do the testing, we could add more platforms. Not sure about hpux, but at least, it's among the supported platforms for Neo4j itself so it makes sense to build it.

manandbytes commented:
@nawroth, thanks for comments

> To merge the PR, we need a CLA

Done. I'd like to squash/rebase commits before merge.

> The packaging needs some work

Will take a look.

nawroth commented:
Ok, confirming that the CLA is in.

Squashing commits is what we usually do.

Just tell us if you have any questions regarding the packaging part. Because of the EPL vs GPL mismatch, including the full license info in the packages is crucial.

manandbytes commented on .travis.yml:
```diff
@@ -0,0 +1,7 @@
+jdk:
```
Not sure if it actually makes sense to build against different JDK, @nawroth WDYT?

nawroth replied:
Unless Neoclipse has some code that requires Java7, I'd say go for oraclejdk6 for 1.9 series and oraclejdk7 for 2.0 series (when we get that far), which is what Neo4j has as it's primary Java platforms.

Don't think building against openjdk makes much sense, when there's no tests to execute! (the code in there is just some small examples, no tests)

manandbytes replied:
The sole point of this CI build is to give a quick feedback that pull requests via Github are in a good shape (i.e., code still compiles and doesn't breaks anything in an obvious way).

Travis CI doesn't have oraclejdk6 out of the box (according to http://about.travis-ci.org/docs/user/languages/java/). However, it should be possible to install it (https://github.com/travis-ci/travis-cookbooks/tree/master/ci_environment/java/recipes) but requires some extra efforts.

Could we use oraclejdk7 right now?

manandbytes replied:
@nawroth Well, Travis CI ignores requested oraclejdk6 and uses Oracle JDK 7u21 instead, https://api.travis-ci.org/jobs/13365085/log.txt?deansi=true

```
$ cd manandbytes/neoclipse
$ git checkout -qf 6176f4a65f7a8435a80beaa77516de7047c9e424
$ jdk_switcher use oraclejdk6
$ java -version
java version "1.7.0_21"
OpenJDK Runtime Environment (IcedTea 2.3.9) (7u21-2.3.9-0ubuntu0.12.04.1)
OpenJDK 64-Bit Server VM (build 23.7-b01, mixed mode)
$ javac -version
javac 1.7.0_21
```

nawroth replied:
Ok, then just go for oraclejdk7.

manandbytes replied:
Done:

```
$ jdk_switcher use oraclejdk7
Switching to Oracle JDK7 (java-7-oracle), JAVA_HOME will be set to /usr/lib/jvm/java-7-oracle
$ java -version
java version "1.7.0_25"
Java(TM) SE Runtime Environment (build 1.7.0_25-b15)
Java HotSpot(TM) 64-Bit Server VM (build 23.25-b01, mixed mode)
$ javac -version
javac 1.7.0_25
```

manandbytes committed:
Extract LICENSE.txt, NOTICE.txt and README.txt from org.neo4j.neoclipse bundle

- during package phase, materialize products with
  tycho-p2-director-plugin's materialize-products goal;
- during package phase, unpack required files from just built
  org.neo4j.neoclipse bundle using Ant's unzip task;
- during pre-integration-test phase (instead of package phase, which is
  default), assemble final archives with tycho-p2-director-plugin's
  archive-products goal.

manandbytes committed:
Place LICENSE.txt, NOTICE.txt and README.txt into every product

Materialized products follow such folder structure:

    > tree -d -L 4 repository/target/products/neoclipse/
    repository/target/products/neoclipse/
    |-- macosx
    |   |-- carbon
    |   |   `-- x86
    |   |       `-- neoclipse-1.9.1-SNAPSHOT
    |   `-- cocoa
    |       |-- x86
    |       |   `-- neoclipse-1.9.1-SNAPSHOT
    |       `-- x86_64
    |           `-- neoclipse-1.9.1-SNAPSHOT
    `-- win32
        `-- win32
            |-- x86
            |   `-- neoclipse-1.9.1-SNAPSHOT
            `-- x86_64
                `-- neoclipse-1.9.1-SNAPSHOT

Place extracted files into every product's root directory with a copy
target (unzip target has no support for multiple mappers).

As there is no way to tell a mapper to place a file into every directory
4 levels deep, we have to enumerate all of them explicitly. To find all
such target directories, run:

    > find repository/target/products/neoclipse/ -mindepth 3 -maxdepth 3 -type d
    repository/target/products/neoclipse/hpux/gtk/ia64_32
    repository/target/products/neoclipse/linux/gtk/x86
    repository/target/products/neoclipse/linux/gtk/ppc64
    repository/target/products/neoclipse/linux/gtk/x86_64
    repository/target/products/neoclipse/win32/win32/x86
    repository/target/products/neoclipse/win32/win32/x86_64
    repository/target/products/neoclipse/macosx/cocoa/x86
    repository/target/products/neoclipse/macosx/cocoa/x86_64
    repository/target/products/neoclipse/macosx/carbon/x86

[1] Mapping File Names, https://ant.apache.org/manual/Types/mapper.html

manandbytes commented:
@nawroth It's almost done, could you take a look?

The last thing to fix is missing icons, like:

```
[INFO] --- tycho-p2-publisher-plugin:0.19.0:publish-products (default-publish-products) @ repository ---
java.io.FileNotFoundException: /home/travis/build/manandbytes/neoclipse/repository/target/products/neoclipse/org.neo4j.neoclipse/icons/obj16/neo-48.xpm (No such file or directory)
```

nawroth commented:
@manandbytes Great work on the packaging!

manandbytes commented:
@nawroth There are still no icons for Windows launchers, due to https://bugs.eclipse.org/bugs/show_bug.cgi?id=384509:

```
[INFO] --- tycho-p2-publisher-plugin:0.19.0:publish-products (default-publish-products) @ repository ---
Error - 7 icon(s) not replaced in /tmp/p2.brandingIron4544998212125010483/launcher.exe using /home/travis/build/manandbytes/neoclipse/repository/target/products/neoclipse/icons/obj16/neo.ico
Error - 7 icon(s) not replaced in /tmp/p2.brandingIron6508182884875601395/launcher.exe using /home/travis/build/manandbytes/neoclipse/repository/target/products/neoclipse/icons/obj16/neo.ico
```

As I understand (but not 100% sure), it requires switching to Kepler (building Kepler-based product).

nawroth commented:
Not worth it trying to fix the launcher icons under windows. At least, the application icon works, so pinning Neoclipse to the taskbar (or whatever it's called) looks perfect in Windows 8, I just tried that.

I suggest this roadmap for the near future of Neoclipse:
- When you squash the commits, we merge this PR.
- Next up is upgrading to 1.9.5. (just released)
- When 1.9.5 is in, I could do a release of Neoclipse and upload the packages. (actually, anyone with upload permissions and maven installed should be able to do this)
- After that's it's time to branch off the 1.9 series (we typically name that branch `1.9-maint`) and start upgrading to the 2.0 series, which will be a larger effort.

WDYT @manandbytes  @peterneubauer @jexp @nrkkaylan ?

manandbytes commented:
> When you squash the commits, we merge this PR.

Not sure if it makes sense to squash them further:

```
> git log --reverse --oneline origin/master..manandbytes/tycho
50cfa19 Enable Travis CI builds
372ef19 Default target platform
e9d0af8 Build with Tycho
52aa3e0 Build product
cf28291 Extract LICENSE.txt, NOTICE.txt and README.txt from org.neo4j.neoclipse bundle
0ddba4f Place LICENSE.txt, NOTICE.txt and README.txt into every product
```

nawroth commented:
Sorry @manandbytes, I didn't notice the commits very already squashed together!

nawroth commented:
@manandbytes Thanks a lot, having CLI builds for Neoclipse is a huge improvement. Awesome you managed to nail so many details at once!
